packages/core/src/tweening/spring.ts

```diff
@@ -0,0 +1,200 @@
+import {decorate, threadable} from '../decorators';
+import {ThreadGenerator} from '../threading';
+import {useProject, useThread, useLogger} from '../utils';
+
+type ProgressFunction = (value: number, time: number) => void;
+
+decorate(spring, threadable());
+export function spring(
+  spring: Spring | null,
+  from: number,
+  to: number,
+  settleTolerance: number,
+  onProgress: ProgressFunction,
+  onEnd?: ProgressFunction,
+): ThreadGenerator;
+
+export function spring(
+  spring: Spring | null,
+  from: number,
+  to: number,
+  onProgress: ProgressFunction,
+  onEnd?: ProgressFunction,
+): ThreadGenerator;
+
+export function* spring(
+  spring: Spring | null,
+  from: number,
+  to: number,
+  settleToleranceOrOnProgress: number | ProgressFunction,
+  onProgressOrOnEnd?: ProgressFunction,
+  onEnd?: ProgressFunction,
+): ThreadGenerator {
+  const settleTolerance =
+    typeof settleToleranceOrOnProgress === 'number'
+      ? settleToleranceOrOnProgress
+      : 0.001;
+
+  onEnd =
+    typeof settleToleranceOrOnProgress === 'number' ? onEnd : onProgressOrOnEnd;
+
+  const onProgress: ProgressFunction = (value: number, time: number) => {
+    if (typeof settleToleranceOrOnProgress === 'function') {
+      settleToleranceOrOnProgress(value, time);
+    } else if (typeof onProgressOrOnEnd === 'function') {
+      onProgressOrOnEnd(value, time);
+    }
+  };
+
+  spring = spring ?? {
+    mass: 0.05,
+    stiffness: 10,
+    damping: 0.5,
+  };
+
+  if (spring.mass <= 0) {
+    useLogger().error(new Error('Spring mass must be greater than 0.'));
+    return;
+  }
+  if (spring.stiffness < 0) {
+    useLogger().error(
+      new Error('Spring stiffness must be greater or equal to 0.'),
+    );
+    return;
+  }
+  if (spring.damping < 0) {
+    useLogger().error(
+      new Error('Spring damping must be greater or equal to 0.'),
+    );
+    return;
+  }
+
+  const project = useProject();
+  const thread = useThread();
+
+  const startTime = thread.time();
+
+  // figure out the step length for each frame
+
+  let position = from;
+  let velocity = spring.initialVelocity ?? 0;
+  let settled = false;
+
+  const update = (dt: number) => {
+    if (spring === null) {
+      return;
+    }
+    const positionDelta = position - to;
+
+    // Using hooks law: F=-kx; with k being the spring constant and x the offset
+    // to the settling position
+    const force = -spring.stiffness * positionDelta - spring.damping * velocity;
+
+    //updating the velocity
```

Comment:
I feel that a lot of these comments don't really add much information. Most of them seem to be simply be describing what the next line does. If it were up to me (which it isn't), I'd say that we should remove these "what" comments and focus more on the "why" comments.

For instance, the comment on line 89 is a good comment because it adds context that isn't immediately obvious by the code itself.

Reply:
Yea, that’s the style I often comment. But I am not a good programmer at all, so i‘ll love to take advices.

Besides that, this whole PR is more of an draft at the moment. So feel free to review again when a final solution is found, I will mention you ;-)

Reply:
I removed some comments or changed them and just kept the comments from jacob‘s changes.